Adjust browser chrome to fit the size of the font (#923)
browserengineering/book: book/chrome.md

class Tab:
    def draw(self, canvas):
        for cmd in self.display_list:
            if cmd.top > self.scroll + HEIGHT - self.chrome_bottom:
                continue
            if cmd.bottom < self.scroll: continue
            cmd.execute(self.scroll - self.chrome_bottom, canvas)
```

Now let's turn our attention to designing the UI.

First, there are still sometimes going to be halves of letters that stick out
into the browser chrome, but we can hide them by just drawing over
them (here I'm assuming we've already computed `self.bottom` (representing
the bottom y coordinate of the browser chrome); that will

> come in a moment):

Reviewer: Let's just move the outer parenthetical to a sentence, I think you're not supposed to do nested parentheses in normal prose :)

``` {.python}
class Chrome:
    def paint(self):
        cmds = []
        cmds.append(
            DrawRect(0, 0, WIDTH, self.bottom, "white"))
        return cmds
```

You'll also need to adjust `scrolldown` to account for the height of

> the page content now being `HEIGHT - self.bottom`:

Reviewer: Hmm, we've been passing `chrome_bottom` to each `Tab`, but in Chapter 15 we instead seem to care about `frame_height`. Should we pass `HEIGHT - self.bottom` instead? It will align more with Chapter 15 and also the code will be a tiny bit nicer.

> the page content now being `HEIGHT - self.bottom`:

Reviewer: Also shouldn't this be `HEIGHT - self.chrome_bottom`?

        )
```

> Now let's figure out the browser chrome size. We'll end up with the following:

Reviewer: ```suggestion
Now let's design the browser chrome. I think it should have the following:
```

* At the top, a list of tab names, separated by vertical lines, and a "`+`"
  button to add a new tab.

* Underneath, the URL of of the current web page, and a "`<`" button to
  represent the browser back-button.

> Let's reserve some space for all this at the top of the window. But how much?

Reviewer: ```suggestion
We'll need reserve some space for all this at the top of the window. But how much?
```

One way could be to pick some arbitrary browser chrome height, then try to
squeeze the chrome into it (and making its font smaller as necessary). Another,
better way, is to pick a font size that is easy enough to read, then compute
the chrome height accordingly.

> Given this design, we can now figure out `self.bottom`: it's the vertical

Reviewer: ```suggestion
For our design, the chrome's height should be the vertical
```

height of those two lines plus some padding between and after them. For
convenience, and to use them later for processing mouse clicks, I'll store

> these parameters on the `Browser` object:^[I also chose `20px` as the

Reviewer: ```suggestion
these parameters on the `Chrome` object:^[I also chose `20px` as the
```

        self.tab_header_bottom = font_height + 2 * self.padding
        self.addressbar_top = self.tab_header_bottom + self.padding
        self.bottom = \
            self.addressbar_top + font_height + \
            2 * self.padding    
```

In addition, it's convenient to define some methods that return the bounds of
the plus and each tab. Note how the size of each
elements' bounds is as wide as the font says it is. That way, if we change our
mind later, we can just change the font size, and everything else just draws
correctly.^[Also, we chose "Tab 1" as the fixed size of a tab, but in real
life many fonts draw different numbers (like "1" and "6") with different
widths, and this approximation also doesn't work for two-digit tab numbers.
Real browsers size their tabs to fit the [title], but we skipped that for
simplicity.]

[title]: https://developer.mozilla.org/en-US/docs/Web/API/Document/title

``` {.python}
class Chrome:
    def plus_bounds(self):
        plus_width = self.font.measure("+")
        return (self.padding, self.padding,
            plus_width + self.padding,
            self.tab_header_bottom - self.padding)

    def tab_bounds(self, i):

>         tab_start_x = self.font.measure("+") + \

Reviewer: Some of these `bounds` methods I'd like to be a little nicer to read. Some options:

- Minimally, can we do `self.padding + self.font.measure("+") + self.padding`, reflecting the actual left-to-right order?
- Or can we call `self.plus_bounds` and do offsets from there?
- I still think we could do better if it was one `layout` method, but I agree we shouldn't delay this PR

        )
```

Now let's design the browser chrome. I think it should have the following:

* At the top, a list of tab names, separated by vertical lines, and a "`+`"
  button to add a new tab.

* Underneath, the URL of of the current web page, and a "`<`" button to
  represent the browser back-button.

> We'll need reserve some space for all this at the top of the window. But how much?

Author: ```suggestion
We'll need to reserve some space for all this at the top of the window. But how much?
```

happen in the `Browser` class, since it's the `Browser` that stores
which tab is active. So let's go to the `handle_click` method and add
a branch for clicking on the browser chrome, which will delegate to a new
method on the `Chrome` object.

``` {.python}
class Browser:
    def handle_click(self, e):
        if e.y < self.chrome.bottom:
            self.chrome.click(e.x, e.y)
        else:
            self.tabs[self.active_tab].click(
                e.x, e.y - self.chrome.bottom)
        self.draw()
```

``` {.python}
class Chrome:
    def click(self, x, y):
        # ...
```

When the user clicks on the browser chrome (the `if` branch), the
browser handles it directly, but if the click is on the page content
(the `else` branch) it is still forwarded to the active tab,
subtracting `self.chrome.bottom` to fix up the coordinates.

If the `y` coordinate is as high as `self.chrome.bottom`, it's just a matter
of comparing it against each of the bounds methods we've already defined. To
that end, let's add a quick method to test whether a point intersects one
of them:

``` {.python}
def intersects(x, y, rect):
    (left, top, right, bottom) = rect

>     return x >= left and x < right and y >= top and y < bottom

Reviewer: This is sort of neither here nor there but this method (and some general chaos in the `DrawXXX` methods) makes me think that we'd be better off defining a simple `Rect` class for the Tk chapters that transitions into `skia.Rect` once we move to Skia.

address bar that shows the current URL would help a lot.

``` {.python}
class Chrome:
    def paint(self):
        # ...
        left_bar = addressbar_left + self.padding
        top_bar = addressbar_top + self.padding
        # ...
            url = str(self.browser.tabs[self.browser.active_tab].url)
            cmds.append(DrawText(
                left_bar,
                top_bar,

>                 url, self.font, "black"))

Reviewer: What's with the indent? If we want to change stuff within a block I always put the first line of the block somewhere to notate that.

should add characters at the cursor. (Remember that the cursor can be
before the first character or after the last!)

*Multiple windows*: Add support for multiple browser windows in
addition to tabs. This will require keeping track of multiple Tk
windows and canvases and grouping tabs by their containing window.
You'll also need some way to create a new window, perhaps with a
keypress such as `Ctrl+N`.

> *Reusing HTML*: Browser chrome is quite complicated in real browsers, and even
> their layouts contain a lot of details, such as font sizes, padding, outlines,

Reviewer: ```suggestion
*Reusing HTML*: Browser chrome is quite complicated in real browsers,
with tricky details such as font sizes, padding, outlines,
```
